### DISCUSSION TITLE
Which hardware is supported?

### DISCUSSION THREAD

**Discussion by niutech** [aduros/wasm4#566] [Questions] [answered]
Will WASM-4 work on ESP32 microcontroller + one of these LCD displays:
- Adafruit 1.3" Color TFT Bonnet (240x240 px)
- Waveshare 1.3" IPS LCD Display HAT (240x240 px)
- ST7789 1.3" IPS TFT LCD Display Module (240x240 px)

or game consoles:
- ODROID-GO
- ESPlay Micro
- WiFiBoy32
- microByte

If you use Wasm3, it should be possible to run WASM-4 games in [Wasm3 supported hardware](https://github.com/wasm3/wasm3/blob/main/docs/Hardware.md), but how to interface LCD displays and physical buttons with WASM-4 API?

**Comment by aduros** [comment, marked as answer]
It should work, and others have had some luck getting it ported to different devices, eg: https://twitter.com/alvaroviebrantz/status/1518343016011943939

If your device is supported by [RetroArch](https://www.retroarch.com/?page=platforms), you should be able to use the libretro WASM-4 core directly.

Otherwise you'll need to write a new [backend](https://github.com/aduros/wasm4/tree/main/runtimes/native/src/backend) that implements the platform-specific I/O parts, and add a new CMake target to build with the new source files. You can check out the existing backends to see what would need to be implemented.

**Reply by aduros** [reply]
btw if you do get something working, please consider opening a PR 😃
